File: lua/tinsert.lua
if #ARGV ~= 4 then
  return redis.error_reply("ERR wrong number of arguments for 'tinsert' command")
end

local child = ARGV[2]
if child == id then
  return redis.error_reply("ERR parent node cannot be same with new node")
end

local childHasChild = redis.call('exists', prefix .. child)
if childHasChild then
  if getPath(child, id) then
    return redis.error_reply("ERR parent node cannot be the posterity of new node")
  end
end

-- Among BEFORE, AFTER and INDEX
local insertType = string.upper(ARGV[3])
local insertPivot = ARGV[4]

local parentHasChildren
local parentChildList = redis.call('get', key)
if parentChildList then
  parentHasChildren = true
  parentChildList = cmsgpack.unpack(parentChildList)
else
  parentHasChildren = false
  parentChildList = {}
end

-- Convert BEFORE and AFTER to INDEX by finding out
-- the index of the pivot
if insertType == 'BEFORE' or insertType == 'AFTER' then
  local index
  for i, v in ipairs(parentChildList) do
    if v[1] == insertPivot then
      index = i
      break
    end
  end

  -- If pivot is not found, set the index to
  -- head (BEFORE) or tail (AFTER)
  if index then
    if insertType == 'BEFORE' then
      insertPivot = index - 1
    else
      insertPivot = index
    end
  else
    if insertType == 'BEFORE' then
      insertPivot = 0
    else
      insertPivot = -1
    end
  end

  insertType = 'INDEX'
else
  insertPivot = tonumber(insertPivot)
end

if insertType ~= 'INDEX' then
  return redis.error_reply('ERR Invalid insert option')
end

local parentChildListLength = #parentChildList

-- Support negative index
if insertPivot < 0 then
  insertPivot = parentChildListLength + insertPivot + 2
else
  insertPivot = insertPivot + 1
end

-- Handle out of range
if insertPivot < 1 then
  insertPivot = 1
elseif insertPivot > parentChildListLength + 1 then
  insertPivot = parentChildListLength + 1
end

table.insert(parentChildList, insertPivot, { child, childHasChild })

-- Update parent childCount
if not parentHasChildren then
  updateHasChildenCache(id, 1)
end

redis.call('set', key, cmsgpack.pack(parentChildList))
redis.call('sadd', prefix .. child .. '::P', id)

-- Return the inserted position
return insertPivot - 1

File: test/tinsert.js
describe('tinsert', function () {
  beforeEach(function () {
    return redis.tinsert('tree', 'ROOT', 'pivot');
  });

  it('inserts a node into a non-exists tree', function () {
    return redis.tinsert('non-exists tree', 'parent', 'node', { index: 0 }).then(function (index) {
      expect(index).to.eql(0);
      return redis.tchildren('non-exists tree', 'parent');
    }).then(function (children) {
      expect(children).to.eql([{ node: 'node', hasChild: false }]);
    });
  });

  describe('INDEX', function () {
    it('inserts to the tail when not specified options', function () {
      return redis.tinsert('tree', 'ROOT', 'node').then(function (index) {
        expect(index).to.eql(1);
        return redis.tchildren('tree', 'ROOT');
      }).then(function (children) {
        expect(children).to.eql([{ node: 'pivot', hasChild: false }, { node: 'node', hasChild: false }]);
      });
    });

    context('when specified', function () {
      beforeEach(function () {
        return redis.tinsert('tree', 'ROOT', 'tail');
      });

      it('inserts to the specified index', function () {
        return redis.tinsert('tree', 'ROOT', 'head', { index: 0 }).then(function (index) {
          expect(index).to.eql(0);
        });
      });

      it('supports negative index', function () {
        return redis.tinsert('tree', 'ROOT', 'middle', { index: -2 }).then(function (index) {
          expect(index).to.eql(1);
        });
      });

      it('insert to the tail when out of range', function () {
        return redis.tinsert('tree', 'ROOT', 'end', { index: 200 }).then(function (index) {
          expect(index).to.eql(2);
        });
      });

      it('insert to the head when out of range (negative)', function () {
        return redis.tinsert('tree', 'ROOT', 'begin', { index: -200 }).then(function (index) {
          expect(index).to.eql(0);
        });
      });
    });
  });

  describe('BEFORE', function () {
    beforeEach(function () {
      return redis.tinsert('tree', 'ROOT', '1');
    });

    it('inserts before the pivot', function () {
      return redis.tinsert('tree', 'ROOT', '0', { before: 1 }).then(function (index) {
        expect(index).to.eql(1);
      });
    });

    it('inserts to the head when pivot is not found', function () {
      return redis.tinsert('tree', 'ROOT', '0', { before: 'non-exists' }).then(function (index) {
        expect(index).to.eql(0);
      });
    });
  });

  describe('AFTER', function () {
    beforeEach(function () {
      return redis.tinsert('tree', 'ROOT', '1');
    });

    it('inserts after the pivot', function () {
      return redis.tinsert('tree', 'ROOT', '0', { after: 'pivot' }).then(function (index) {
        expect(index).to.eql(1);
      });
    });

    it('inserts to the tail when pivot is not found', function () {
      return redis.tinsert('tree', 'ROOT', '0', { after: 'non-exists' }).then(function (index) {
        expect(index).to.eql(2);
      });
    });
  });
});
